File: 2026/08-August/21-Friday/solution.py
from typing import List

class Solution:
    def findKthSmallest(self, coins: List[int], k: int) -> int:
        from math import gcd

        def lcm(a, b):
            return a // gcd(a, b) * b

        def count(x):
            # Count numbers <= x divisible by at least one coin
            n = len(coins)
            total = 0

            # Inclusion-exclusion
            for mask in range(1, 1 << n):
                multiple = 1
                bits = 0
                valid = True

                for i in range(n):
                    if mask & (1 << i):
                        bits += 1
                        multiple = lcm(multiple, coins[i])

                        if multiple > x:
                            valid = False
                            break

                if not valid:
                    continue

                cnt = x // multiple

                if bits % 2 == 1:
                    total += cnt
                else:
                    total -= cnt

            return total

        left = 1
        right = min(coins) * k

        while left < right:
            mid = (left + right) // 2

            if count(mid) >= k:
                right = mid
            else:
                left = mid + 1

        return left

File: 2026/08-August/21-Friday/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long findKthSmallest(vector<int>& coins, int k) {
        int n = coins.size();

        auto lcm = [](long long a, long long b) {
            return a / gcd(a, b) * b;
        };

        auto count = [&](long long x) {
            long long total = 0;

            for (int mask = 1; mask < (1 << n); mask++) {
                long long multiple = 1;
                int bits = 0;
                bool valid = true;

                for (int i = 0; i < n; i++) {
                    if (mask & (1 << i)) {
                        bits++;

                        multiple = lcm(multiple, (long long)coins[i]);

                        if (multiple > x) {
                            valid = false;
                            break;
                        }
                    }
                }

                if (!valid) continue;

                long long cnt = x / multiple;

                if (bits & 1)
                    total += cnt;
                else
                    total -= cnt;
            }

            return total;
        };

        long long left = 1;
        long long right = 1LL * (*min_element(coins.begin(), coins.end())) * k;

        while (left < right) {
            long long mid = left + (right - left) / 2;

            if (count(mid) >= k)
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
};
